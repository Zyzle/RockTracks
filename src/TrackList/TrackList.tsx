import React, { PureComponent } from 'react';
import { Link } from 'react-router-dom';

import { SearchResults } from '../models/ITunes';

import './TrackList.scss';

interface TrackState {
  results: SearchResults | null;
}

class TrackList extends PureComponent<{}, TrackState> {
  constructor(props: any) {
    super(props);

    this.state = {
      results: null
    };
  }
 
  componentDidMount() {
    fetch('https://itunes.apple.com/search?term=rock&media=music')
      .then((response: Response) => {
        return response.json();
      })
      .then((results: SearchResults) => {
        this.setState({results});
      });
  }
  
  render() {
    let content = <b>No results to display</b>;

    if (this.state.results && this.state.results.results) {
      const results = this.state.results.results.map((result) => {
        return (
          <div key={result.trackId}>
            <img src={result.artworkUrl30} alt={result.trackName}></img>
            <div>
              <div><Link to={`/${result.trackId}`}>{result.trackName}</Link></div> 
              <div>{result.artistName}</div>
              <div>{result.trackPrice}</div>
            </div>
          </div>
        );
      });
      content = <div>{results}</div>;
    }

    return (
      <div>
        {content}
      </div>
    )
  }
}

export default TrackList;