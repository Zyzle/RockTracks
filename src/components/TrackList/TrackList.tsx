import React, { PureComponent } from 'react';
import { Link, RouteComponentProps } from 'react-router-dom';

import { Result} from '../../models/ITunes';
import { TracksState } from '../../reducer';

import './TrackList.scss';

// search function isn't actually used here anymore but will
// leave as an example of how dispatch props could be added 
// to the intersection
export type TrackListProps = TracksState & RouteComponentProps & {
  search :() => void
};

class TrackList extends PureComponent<TrackListProps> {
  constructor(props: TrackListProps) {
    super(props);
  }
  
  render() {
    let content = <b>No results to display</b>;

    if (this.props.results && this.props.results.length) {
      const results = this.props.results.map((result: Result) => {
        return (
          <div key={result.trackId}>
            <img src={result.artworkUrl30} alt={result.trackName}></img>
            <div>
              <div>{result.trackName}</div> 
              <div>{result.artistName}</div>
              <div>${result.trackPrice}</div>
            </div>
            <Link to={`/${result.trackId}`}>
              <button>track info</button>
            </Link>
            <hr />
          </div>
        );
      });
      content = <div>{results}</div>;
    }

    return (
      <div>
        <div>{this.props.pending ? 'Loading....' : ''}</div>
        {content}
      </div>
    )
  }
}

export default TrackList;