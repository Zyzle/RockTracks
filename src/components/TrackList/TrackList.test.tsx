import React from 'react';
import { MemoryRouter } from 'react-router-dom';

import renderer from 'react-test-renderer';

import TrackList, { TrackListProps } from './TrackList';

describe('TrackList Component', () => {

  it('should render', () => {
    const props: TrackListProps = {
      search: jest.fn(),
      history: jest.fn() as any,
    location: jest.fn() as any,
      match: jest.fn() as any,
      resultsCount: 1,
      results: [{
        artistId: 123,
        collectionId: 456,
        trackId: 789,
        artistName: 'An artist',
        collectionName: 'A collection',
        trackName: 'A track',
        trackViewUrl: 'https://www.google.com',
        artworkUrl100: 'https://via.placeholder.com/100',
        artworkUrl30: 'https://via.placeholder.com/30',
        trackPrice: 1.99,
        releaseDate: '01-01-1970',
        trackTimeMillis: 80000,
        currency: 'GBP',
      }],
      pending: false,
      error: null
    };

    const component = renderer.create(
      <MemoryRouter>
        <TrackList {...props} />
      </MemoryRouter>
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});