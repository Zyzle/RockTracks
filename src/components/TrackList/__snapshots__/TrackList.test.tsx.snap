// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrackList Component should render 1`] = `
<div>
  <div>
    
  </div>
  <div>
    <div>
      <img
        alt="A track"
        src="https://via.placeholder.com/30"
      />
      <div>
        <div>
          A track
        </div>
        <div>
          An artist
        </div>
        <div>
          $
          1.99
        </div>
      </div>
      <a
        href="/789"
        onClick={[Function]}
      >
        <button>
          track info
        </button>
      </a>
      <hr />
    </div>
  </div>
</div>
`;
